nixos/modules/services/x11/window-managers/xmonad.nix

```
@@ -43,7 +43,7 @@ in {
       haskellPackages = mkOption {
         default = pkgs.haskellPackages;
         defaultText = "pkgs.haskellPackages";
-        type = types.package;
+        type = types.functionTo (types.listOf types.package);
```

It isn't a function.

I don't think we can do better than this without compromising performance.

```suggestion
        # unspecified: Avoid the cost of evaluating _every package_ in the `haskellPackages` attrset.
        type = types.attrsOf types.unspecified;
```

This is lazy enough, because:

```
nix-repl> (lib.types.attrsOf lib.types.unspecified).check { a = builtins.throw "too strict"; } 
true
```

I suppose we could set a `_type = "haskellPackages";` attribute in those package sets so we can catch the mistake without the cost, but let's fix this first.

---

done

---

I've been wanting to add a `types.primitive "set"`-like type, which would just ensure that `builtins.typeOf <definition> == "set"`, but otherwise not recurse any further. This could be used here.

Another alternative would be to introduce `types.rawValue`, which wouldn't do any checks on the definitions. Then `types.lazyAttrsOf types.rawValue` could be used here.

---

@roberth Oh actually are you sure `attrsOf unspecified` is okay here? I think it would still evaluate all derivations since `attrsOf` is strict in values.

---

@Infinisil I guess it's only strict in attrs when merging. I only checked `check`.

```
nix-repl> (lib.types.attrsOf lib.types.unspecified).check { a = builtins.throw "too strict"; } 
true
```